app/src/main/java/me/phh/sip/SipMessage.kt

```diff
@@ -0,0 +1,119 @@
+package me.phh.sip
+
+import java.io.BufferedReader
+
+
+/* type definitions */
+
+enum class SipMethod {
+	REGISTER,
+	SUBSCRIBE,
+	INVITE,
+	ACK,
+	CANCEL,
+	BYE,
+	OPTIONS,
+	MESSAGE
+}
+
+data class SipStatusCode(val code: Int)
+
+data class SipHeader(val value: String, val parameters: Map<String, String>)
+
+sealed class SipMessage()
+
+data class SipCommonMessage(
+	val firstLine: String,
+	val headers: Map<String, List<SipHeader>>,
+	// TODO: messageBody needs to handle multipart, including nested multipart,
+	// so needs to be a tree of sort? We can probably flatten it into
+	// Map<String, List<ByteArray>> where first string is content type?
+	// (second part list because there could be multiple of the same
+	// and map doesn't allow duplicates)
+	// See RFC5621 section 3
+	val body: ByteArray?,
+)
+
+data class SipRequest(
+	val method: SipMethod,
+	val message: SipCommonMessage,
+) : SipMessage()
+
+data class SipResponse(
+	val statusCode: SipStatusCode,
+	val message: SipCommonMessage,
+) : SipMessage()
+
+
+/* rfc3261 section 7 describes how headers should be formed, in particular:
+ *  - spaces around : are ignored
+ *  - multilines headers are allowed, continuation lines start with a space or tab
+ *    and should be joined with a single space
+ *  - some headers are comma separated lists, multiple occurence of these headers
+ *    is identical to having separated values on single line
+ *  - some headers have parameters split by ;, thanksfully no comma separated list
+ *    has these so there is no ambiguity vs. how they'd distribute in this case...
+ *  - headers and parameters are case-insensitive
+ */
+private val splitHeader = "^(.+?) *: *(.+)$".toRegex()
+
+fun parseHeaders(input: BufferedReader): Map<String, List<SipHeader>> =
+	input.lineSequence()
+  		.takeWhile { it.isNotBlank() }
+		.fold(emptyMap<String, List<SipHeader>>(), fold@{ headers, line ->
+			// ignore anything we don't recognize
+			// TODO: if line starts with a space, content should be happened to previous value
```

Comment:
that's sadly a requirement for me, I have services who send multi-line Authorization header...

Reply:
Ok. That's a bit tricky with the current line by line code, it'd be easier if we do the split ourselves so e.g. we split on \r\n but not \r\n<space> for this function -- I think it's the cleanest thing to do to handle ByteArrays in the messages anyway so I'll do that next :/